src/Compilers/CSharp/Portable/Binder/IndexerArgumentInfo.cs

```diff
@@ -0,0 +1,38 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    /// <summary>
+    /// Information about the arguments of an implicit or explicit Index or Range indexer pattern used in list patterns.
+    /// </summary>
+    internal sealed class IndexerArgumentInfo
+    {
+        public readonly Symbol Symbol;
+        public readonly ImmutableArray<BoundExpression> ArgumentsOpt;
+        public readonly bool Expanded;
+
+        public IndexerArgumentInfo(Symbol symbol, ImmutableArray<BoundExpression> argumentsOpt, bool expanded)
+        {
+            Debug.Assert(
+                symbol is PropertySymbol { IsIndexer: true } ||
+                symbol is MethodSymbol { Name: WellKnownMemberNames.SliceMethodName or nameof(string.Substring) } && argumentsOpt.IsDefault);
+
+            Symbol = symbol;
+            ArgumentsOpt = argumentsOpt;
+            Expanded = expanded;
+        }
+
+        public static IndexerArgumentInfo CreateImplicitIndexOrRangeIndexerInfo(Symbol symbol)
+        {
+            return new IndexerArgumentInfo(symbol, argumentsOpt: default, expanded: default);
```

Comment:
>, argumentsOpt: default, expanded: default [](http://example.com/codeflow?start=49&length=42)

My point wasn't to move this into a separate method: my point was that this isn't safe. Initial binding always needs to calculate these values. I don't think we should have this class at all: we should just be using the appropriate bound node and going through the existing paths for binding indexer accesses, rather than trying to duplicate lots of logic around and then having multiple places that need to be fixed up.

Reply:
When these are `default` we have an implicit range/index support. e.g. `this[int]` or `Slice(int, int)` etc, which is impossible to have params or default args. Note that this is later used in lowering only.

In the previous iteration where I used BindIndexer, we had to have two field BoundIndexer+Symbol in the bound node.
With IndexerArgumentInfo we no longer need it - that remains true even if we go with BindIndexer for the lookup.

Reply:
> In the previous iteration where I used BindIndexer, we had to have two field BoundIndexer+Symbol in the bound node.

I don't know why we'd need that, but if that's what we need then that's what we need. I don't think we should have this class.

Reply:
>I don't know why we'd need that,

Because we don't run overload resolution for that. See TryBindIndexOrRangeIndexer. We directly fetch the symbol. There's no other info to capture.

Reply:
I'd agree the duplication seems unnecessary. I'll follow PerformPatternMethodLookup which I think is closer to what you have in mind.